### ISSUE TITLE
SSR: How to protect oneself from empty objects

### ISSUE THREAD

**Issue by bmarkovic** [nuxt/nuxt.js#2632]
I have an object passed down with a `prop` from a page component to a regular component. Sometimes the object is undefined (depending on an `asyncData` handled pre-fetch of different object from the backend) before it's subsequently populated with Ajax calls.

So I have something like this in template:
```html
<span v-if="thing">{{ (thing || {}).name || 'Dis ting got no name' }}</span>
```
Note that I **currently** have both checks (i.e. `v-if` and `thing || {}`), either is good enough for Vue when `thing` is undefined.

Neither stops Nuxt from crashing with `Cannot read property 'name' of undefined` if I reload the page i.e. when the whole page is first rendered server-side despite the fact that:

* in both Node and in ES5+ compliant browser javascript, as long as `thing` is defined (in the sense that the lexeme exists in the scope) the syntax `(thing || {}).name` will return either thing.name or undefined, and if the lexeme doesn't exist in the scope a different error, `ReferenceError: thing is not defined` would be thrown)
* The very existence of `v-if=" ... "` should prevent the template interpolation to be evaluated at all if the `v-if` condition evaluates to `flase`.

This creates a significant (in my opinion) discrepancy between how Vue and Vue SSR under Nuxt behave which I don't see covered in any of the documentation.

**Comment by Mikeysax** [comment]
Hey @bmarkovic  I think the main issue might be the logic you are using more than a fault of SSR or Nuxt.

Empty Objects are considered truthy, so logic like `if { }` will return true because the Object does exist. When dealing with Objects, you can check the existance of it but you also need to check if it's populated.

One way of doing this is by grabbing the keys and determining if the length is greater than 0:
```js
Object.keys(thing).length > 0;
```
If you're using Vuex, you can set a getter that does this or create a helper you can easily import into different components. 

I hope this help!

**Comment by Wei12345** [comment]
Usually I use  `computed` or `fetch` or `asynData`to initialize or preset it
Avoid it to appear `undefined `or `null `or `empty string`

**Comment by bmarkovic** [comment]
@Wei12345 I do use `asyncData` in the page component (where it's only available), however this component is not page component but a component under the page component that gets the data passed downstream from a page component with a `prop` binding. However, I **need** this object to possibly be empty and templates **should** allow me to test for it and, still, it's passed down from the page component as a `prop`, so it's a "defined" identifier, just equal to `undefined`

Having little experience with Nuxt/Vue-SSR (but I consider myself experienced in client-side Vue(x)) I also doubt `computed` will work server-side as there is "no reactivity in server-side rendering, because (performance) reasons"  -- Evan's (not exact) words -- what precisely did he mean by that, he didn't really elaborate and I'm not entirely sure TBH.

But one of the things it, sadly, means is that `v-if` **doesn't work at all** (see below)

@Mikeysax Unfortunately your suggestion of testing for `Object.key(thing).length` doesn't work, which I found out in a "roundabout" fashion (see my workaround below). With my old code it doesn't stop the template interpolation to be evaluated as `undefined.name`, and as I said, it's probably due to reactivity limitations in SSR where `v-if` doesn't actually remove "tags" from the VDOM. To reiterate, thing wasn't `{}` in that case (which wouldn't be a problem) but `undefined` which is why my logic for `v-if` worked in Vue (client-side).

So far the workaround is having this in `asyncData`

```javascript
async asyncData({store, route}) {
  if (process.server) {
    thong = await store.getters.fetchThong(route.params.id)
    thing = (thong && thong.thing_id) && await store.getters.fetchThing(thong.thing_id)
  } else {
    // client side logic has caching and stuff I can't be bothered to copy and isn't 
    // really relevant
  }
  thong = thong || {}
  thing = thing || {}
  return { thong, thing }
}
```
However the `v-if="Object.keys(thing).length > 0"` doesn't really work in SSR so I have the entire block under `v-if` still rendered despite it being an empty object. If it's an `undefined` object -- `Cannot read property 'name' of undefined` rears it's ugly head.

Needless to say, I'm not terribly happy about it :)

**Comment by Wei12345** [comment]
@bmarkovic 
I also encountered similar problems
Currently I am doing this
```
<span v-if="thing.name">{{ thing.name }}</span>
<span v-else>Dis ting got no name</span>
```
Have you tried this before?
```
<span v-if="thing && thing.name">{{ thing.name }}</span>
<span v-else>Dis ting got no name</span>
```
I'm not sure if this is feasible
Not sure if this is what you want?

**Comment by bmarkovic** [comment]
@Wei12345 I wasn't able to play with this project until now. It appears that your suggestion does actually work, in both SSR and Vue consistently. I do wonder why does it appear to try to evaluate contents of `v-if` even if the condition isn't met, whereas if `v-else` is present it doesn't is beyond me but thank you.

**Comment by qm3ster** [comment]
@bmarkovic currently, does `v-if="undefined"` ever lead to the contents being evaluated?

**Comment by bmarkovic** [comment]
@qm3ster I have upgraded my Nuxt project using `ncu` to be sure we're all on the same page.

I have removed all "safety nets" from the project (thong = thong || {}) and resorted to v-if="" checks in the tamplate as this:

```html
<span
  v-if="thing && Object.keys(thing).length > 0"
  v-text="' ' + (thing.name || 'Dis ting got no name')
>
```
It appears not to happen now. I do have other issues when I remove safeguards: When loading that page over SSR and `thing` is undefined, I get [Vue warn] that `thing` is not defined but referenced (despite it being passed by props, just not successfully fetched in `asyncData()` of the page component).
